### PR TITLE
Dexter lifecycle sample

### PR DESCRIPTION
**Pull request by flipper83**
- change the main screen for chooser screen where you can select the sample to run.
- add sample that test request permission in onCreate.
- add sample that test request on application initialize.

I've found that if you request Dexter from application and you use a Dialog, the dialog crash because the window doesn't exisit. We need talk about witch solution we are going to give to this one.

### CODE REVIEW

File: sample/src/main/java/com/karumi/dexter/sample/ChooserActivity.java [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+    OnCreateSampleActivity.open(this);
+  }
+
+  @OnClick(R.id.bt_application_sample)
+  public void openApplicationSample() {
+    DexterApplicationMode.enable(this);
+    relaunchApplicationToSample();
+  }
+
+  private void relaunchApplicationToSample() {
```
Comment (flipper83) [inline sample/src/main/java/com/karumi/dexter/sample/ChooserActivity.java]: The method from this code is not the most beautiful sample, but I thought that it's de easy way to test this one without create two sample project.

File: sample/src/main/java/com/karumi/dexter/sample/SampleApplication.java [hunk 2]
```diff
@@ -27,5 +29,14 @@
...
     Dexter.initialize(this);
+    boolean launchDexterOnApplication = DexterApplicationMode.isEnable(this);
+    if (launchDexterOnApplication) {
+      DexterApplicationMode.clear(this);
+      requestDexterOnApplication();
+    }
+  }
+
+  private void requestDexterOnApplication() {
+    Dexter.checkPermission(new EmptyPermissionListener(), Manifest.permission.RECORD_AUDIO);
```
Comment (flipper83) [inline sample/src/main/java/com/karumi/dexter/sample/SampleApplication.java]: I'm not using any kind of listener, do you wish that I store any info in a singleton to read from the activity sample to show any kind of feedback ??

**Comment by Serchinastico** [comment]
I'd personally keep this sample separate because it's meant for testing purposes and not for educational ones.

Besides, these are the things I've seen while manual testing the samples:
* In the `OnCreateSampleActivity`, when the user accepts the permission for the first time everything works like a charm. That means I'm seeing the expected flow `onCreate` -> `onResume` -> `onPause` -> Permission dialog -> `onResume`. However, if I accept the permission, close the app and go to the activity again, the flow is exactly the same (obviously with no dialog involved). There is no reason to have an `onResume` -> `onPause` -> `onResume` transition. When we attack the #47 issue this can be a great starting point for testing the problem.
* Same issue happens in the `ApplicationSampleActivity`.
* No weird transitions when calling Dexter from `Activity.onCreate` whatsoever, great.
* No weird transitions when calling Dexter from `Application.onCreate`, 👌 .

File: sample/src/main/java/com/karumi/dexter/sample/ChooserActivity.java [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+import android.os.Bundle;
+import butterknife.ButterKnife;
+import butterknife.OnClick;
+
+/**
+ * Activity to choose applications options.
+ */
+public class ChooserActivity extends Activity {
+
+  private static final int TIME_TO_RESTART = 50;
```
Comment (Serchinastico) [inline sample/src/main/java/com/karumi/dexter/sample/ChooserActivity.java]: I found myself using a smaller time, 1 or 2 seconds is ok

File: sample/src/main/java/com/karumi/dexter/sample/ChooserActivity.java [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+    ButterKnife.bind(this);
+  }
+
+  @OnClick(R.id.bt_common_sample)
+  public void openCommonSample() {
+    SampleActivity.open(this);
+  }
+
+  @OnClick(R.id.bt_oncreate_sample)
+  public void openOncreateSample() {
```
Comment (Serchinastico) [inline sample/src/main/java/com/karumi/dexter/sample/ChooserActivity.java]: openOn**C**reateSample

File: sample/src/main/java/com/karumi/dexter/sample/ChooserActivity.java (hunk 1 above)
Comment (Serchinastico) [inline sample/src/main/java/com/karumi/dexter/sample/ChooserActivity.java]: It's great as far as it lets us test what's happening

File: sample/src/main/java/com/karumi/dexter/sample/DexterApplicationMode.java [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+
+package com.karumi.dexter.sample;
+
+import android.content.Context;
+import android.content.SharedPreferences;
+
+public class DexterApplicationMode {
+
+  public static final String APPLICATION_SETTING = "applicationSetting";
+  private static final String IS_ENABLE = "IS_ENABLE";
```
Comment (Serchinastico) [inline sample/src/main/java/com/karumi/dexter/sample/DexterApplicationMode.java]: IS_ENABLE**D**

Same for the get method

File: sample/src/main/java/com/karumi/dexter/sample/OnCreateSampleActivity.java [hunk 6]
```diff
@@ -0,0 +1,96 @@
...
+  private PermissionListener feedbackViewPermissionListener = new PermissionListener() {
+    @Override public void onPermissionGranted(PermissionGrantedResponse response) {
+      setFeedback(getString(R.string.permission_granted_feedback));
+    }
+
+    @Override public void onPermissionDenied(PermissionDeniedResponse response) {
+      setFeedback(getString(R.string.permission_denied_feedback));
+    }
+
+    @Override public void onPermissionRationaleShouldBeShown(PermissionRequest permission,
```
Comment (Serchinastico) [inline sample/src/main/java/com/karumi/dexter/sample/OnCreateSampleActivity.java]: We should be calling the token with `token.continuePermissionRequest();`

File: sample/src/main/java/com/karumi/dexter/sample/SampleApplication.java (hunk 2 above)
Comment (Serchinastico) [inline sample/src/main/java/com/karumi/dexter/sample/SampleApplication.java]: I'd use one with the `token.continuePermissionRequest();` method called in the `onPermissionRationaleShouldBeShown` callback.

File: sample/src/main/res/values/strings.xml [hunk 7]
```diff
@@ -24,12 +24,31 @@
   <string name="permission_denied_feedback">DENIED</string>
   <string name="permission_permanently_denied_feedback">PERMANENTLY DENIED</string>
   <string name="permission_unknown_feedback">UNKNOWN</string>
-  <string name="all_permissions_denied_feedback">All those permissions are needed for doing some magic</string>
-  <string name="camera_permission_denied_feedback">Camera permission is needed for doing some magic</string>
-  <string name="contacts_permission_denied_feedback">Contacts permission is needed for doing some magic</string>
+  <string name="all_permissions_denied_feedback">All those permissions are needed for doing some
```
Comment (Serchinastico) [inline sample/src/main/res/values/strings.xml]: Why are we limiting the column size of a configuration file?

File: dexter/src/main/AndroidManifest.xml [hunk 8]
```diff
@@ -16,6 +16,7 @@
   -->
 <manifest xmlns:android="http://schemas.android.com/apk/res/android"
     package="com.karumi.dexter"
+    android:allowBackup="true"
```
Comment (pedrovgs) [inline dexter/src/main/AndroidManifest.xml]: Isn't this the default value for ``allowBackup``? Should we add this param to the library manifest? Couldn't this be a problem during the merge manifest process?

File: sample/src/main/java/com/karumi/dexter/sample/ApplicationSampleActivity.java [hunk 9]
```diff
@@ -0,0 +1,34 @@
...
+
+/**
+ * Activity to show if Dexter works when have been invoked from the application.
+ */
+public class ApplicationSampleActivity extends Activity {
+
+  @Override protected void onCreate(Bundle savedInstanceState) {
+    super.onCreate(savedInstanceState);
+    setContentView(R.layout.application_sample_activity);
+    ButterKnife.bind(this);
```
Comment (pedrovgs) [inline sample/src/main/java/com/karumi/dexter/sample/ApplicationSampleActivity.java]: Do you need ``ButterKnife`` here?

Comment (pedrovgs) [inline sample/src/main/java/com/karumi/dexter/sample/ApplicationSampleActivity.java]: I don't understand the usage of this Activity. Can you provide us more information?

File: sample/src/main/java/com/karumi/dexter/sample/DexterApplicationMode.java [hunk 10]
```diff
@@ -0,0 +1,47 @@
...
+
+  public static boolean isEnable(Context context) {
```
Comment (pedrovgs) [inline sample/src/main/java/com/karumi/dexter/sample/DexterApplicationMode.java]: isEnable**d**

File: sample/src/main/res/layout/application_sample_activity.xml [hunk 11]
```diff
@@ -0,0 +1,36 @@
+<?xml version="1.0" encoding="utf-8"?>
+<!--
+  ~ Copyright (C) 2015 Karumi.
```
Comment (pedrovgs) [inline sample/src/main/res/layout/application_sample_activity.xml]: The rest of the copyright used is from 2016 :S

**Comment by pedrovgs** [comment]
Definitively we should not allow the usage of ``Dexter`` from the ``Application`` object.

**Comment by flipper83** [comment]
Do you want that I tryed launch an exception if somebody call to dexter and windows is not attached yet??

**Comment by pedrovgs** [comment]
Yeap, and we should remove Application example :) What do you think @Serchinastico ?

**Pull request closed without merging**